nanofi/src/processors/file_input.c

```diff
@@ -0,0 +1,406 @@
+    return prepare_message(data, len, as);
+}
+
+void enqueue_chunk(file_input_context_t * ctx, data_buff_t chunk) {
+    if (chunk.len > 0 && chunk.data) {
+        size_t fp_len = strlen(ctx->file_path);
+        chunk.file_path = (char *)malloc(fp_len + 1);
+        strcpy(chunk.file_path, ctx->file_path);
+
+        int length = snprintf(NULL, 0, "%llu", ctx->current_offset);
+        char * offset_str = (char *)malloc(length + 1);
+        snprintf(offset_str, length + 1, "%llu", ctx->current_offset);
+
+        properties_t * props = NULL;
+        add_property(&props, "current offset", offset_str);
+        add_property(&props, "file path", chunk.file_path);
+
+        message_t * msg = create_message(chunk.data, chunk.len, props);
+        free(chunk.data);
+        free(offset_str);
+        free(chunk.file_path);
+        free_properties(props);
+        size_t bytes_enqueued = enqueue_message(ctx->msg_queue, msg);
+        if (bytes_enqueued < chunk.len) {
+            //we were not able to enqueue all of chunk.len bytes
+            //therefore, update the current offset
+            ctx->current_offset -= (chunk.len - bytes_enqueued);
+        }
+    }
+}
+
+void read_file_chunk(file_input_context_t * ctx) {
+    errno = 0;
+    FILE * fp = fopen(ctx->file_path, "rb");
+    if (!fp) {
+        printf("Error opening file %s, error: %s\n", ctx->file_path, strerror(errno));
+        return;
+    }
+    size_t bytes_read = 0;
+    char * data = (char *)malloc((ctx->chunk_size + 1) * sizeof(char));
+    fseek(fp, ctx->current_offset, SEEK_SET);
+    while ((bytes_read = fread(data, 1, ctx->chunk_size, fp)) > 0) {
+        if (bytes_read < ctx->chunk_size) {
+            break;
+        }
+        data[ctx->chunk_size] = '\0';
+        data_buff_t buff;
+        buff.data = (char *)malloc(strlen(data) + 1);
+        memcpy(buff.data, data, strlen(data) + 1);
+        buff.len = strlen(data);
+        ctx->current_offset = ftell(fp);
+        size_t old_offset = ctx->current_offset;
+        enqueue_chunk(ctx, buff);
+        if (old_offset > ctx->current_offset) {
+            //we were not able to ship all of the chunk read from
+            //the file. let's stop reading and come back later
+            break;
+        }
+        fseek(fp, ctx->current_offset, SEEK_SET);
+    }
+    free(data);
+    fclose(fp);
+}
+
+
+void read_file_delim(file_input_context_t * ctx) {
+    FILE * fp = fopen(ctx->file_path, "rb");
+    errno = 0;
+    if (!fp) {
+        printf("Unable to open file. {file: %s, reason: %s}\n", ctx->file_path, strerror(errno));
+        return;
+    }
+
+    char data[4096 + 1];
+    memset(data, 0, 4097);
+    fseek(fp, ctx->current_offset, SEEK_SET);
+    size_t bytes_read = 0;
+    while ((bytes_read = fread(data, 1, 4096, fp)) > 0) {
+        data[bytes_read] = '\0';
+        const char * begin = data;
+        const char * end = NULL;
+
+        while ((end = strchr(begin, ctx->delimiter))) {
```

Comment:
If there is a `\0` in this file (which I think is valid, we could read a delimited binary file), you will exit on it.
I suggest you don't treat the buffer as a C string and use `memchr` instead of `strchr`.

Reply:
fair point. Fixed it.